### PR TITLE
Add a way to easily convert an AndroidAppWaker into a Waker

### PR DESCRIPTION
**Pull request by notgull**
This commit adds a "waker()" method to AndroidAppWaker. It converts it
into an `std::task::Waker`, which is the type of waker used by
asynchronous tasks for scheduling. The goal is to allow AndroidAppWaker
to be easily used to set up an asynchronous context.

The implementation is very efficient, owing to the "static pointer"
style of coding already used. The "wake" function just calls
ALooper_wake, and cloning/dropping the waker is just a copy.

Discussion questions:

- Should `waker()` take `&self` or `self`? I chose the latter.

### CODE REVIEW

**Comment by notgull** [comment]
Build failure is due to codebase rot

**Comment by MarijnS95** [comment]
https://github.com/rust-mobile/android-activity/pull/164#issuecomment-2327429638

Yeah, we're really lacking maintenance here, unfortunately. As always, I'm happy to submit PRs to tackle issues and implement features every once in a while in between other tasks, but **need a second maintainer for review and approval**.

@rib can you update us on your availability and commitment towards this crate, or should we seek for a third maintainer to fill this review-gap?

**Comment by MarijnS95** [comment]
Regarding the PR itself, note that the result of a looper wake is very loosely defined and not to be relied on: #170

If there's any explicit `async` task spawning and progress monitoring that needs to happen on an `ALooper`, I'd prefer if we add an abstraction based on file descriptors or callbacks which, unlike waking, can be explicitly handled.

**Comment by notgull** [comment]
Interesting. Does it have a chance of spuriously not waking up the event loop? `winit` currently assumes it always does.

**Comment by MarijnS95** [comment]
@notgull according to their documentation, it would get batched up with other events, such that the event loop is "active" / "awake" but without ever seeing it return `WAKE` because it got "folded into" the other events.

`winit` currently relies on `Poll::Wake` to trigger a user-requested redraw:

https://github.com/rust-windowing/winit/blob/ae4c449670674d8ac0d6d8754caf3fe5f4954c25/src/platform_impl/android/mod.rs#L586-L602

But it also checked these before entering a `poll_events()`: https://github.com/rust-windowing/winit/blob/ae4c449670674d8ac0d6d8754caf3fe5f4954c25/src/platform_impl/android/mod.rs#L562-L565

With these findings, it'll miss events on `redraw_flag` and `proxy_wake_up`, unless we hoist those checks out of the `match` block for any `Poll`/`Err` returned out of `poll_events()`. Or if we want to be more specific, attach a pipe or eventfd to be guaranteed a consistent result whenever our channel into the `Looper` signaled anything.

**Comment by notgull** [comment]
I would prefer to just use eventfd.

**Comment by MarijnS95** [comment]
Same preference here, it's much more explicit than pretending every `Poll` could have a `Wake` folded into it.

**Comment by rib** [comment]
Sorry for only just coming around to look at this.

This seems reasonable.

Re: #170 and the way that wake events are documented: the point is just that when `_wake()` is used to wake up a Looper then although it does guarantee a wake up (you don't have to worry about it not waking) it's not guaranteed that you _specifically_ get a `::Wake` event if there are other events that can be delivered after the wake up.

Regarding how winit handles `Wake` events, winit is doing the right thing already I believe.

Winit never _specifically_ requires a `::Wake` event for anything - it's kind of the opposite and it will _ignore_ `::Wake` events when it currently has nothing it needs to do.

In itself a `Wake` event is the the last-resort event that's delivered after the even loop wakes after calling `epoll_wait` if there is _nothing_ else to report (such as a timeout or an idle callback or some other file descriptor event).

So from Winit's point of view a Wake event on its own doesn't represent anything interesting that needs to be handled, unless it knows that it requested a wake up.

The android backend for Winit effectively tracks when it's waiting for a wake up and so this code in Winit:

```rust
                android_activity::PollEvent::Wake => {
                    // In the X11 backend it's noted that too many false-positive wake ups
                    // would cause the event loop to run continuously. They handle this by
                    // re-checking for pending events (assuming they cover all
                    // valid reasons for a wake up).
                    //
                    // For now, user_events and redraw_requests are the only reasons to expect
                    // a wake up here so we can ignore the wake up if there are no events/requests.
                    // We also ignore wake ups while suspended.
                    self.pending_redraw |= self.redraw_flag.get_and_reset();
                    if !self.running
                        || (!self.pending_redraw && !self.user_events_receiver.has_incoming())
                    {
                        return;
                    }
                }
```
is about ignoring Wake events when Winit has nothing it needs to do.

In other words, an `ALooper_wake()` may lead to any event (it's only promising that `epoll_wait` or `_pollOnce` will unblock and return with _an_ event in a finite amount of time. This is why the Android NDK docs clarify that you should treat any event like a `Wake` event.  _I suppose in the past they made the mistake of implying in their docs that `_wake()` will always result in a `Wake` which is not the case_.

So in Winit specifically, it only has that special case for ignoring some Wake events and then apart from that, all events (including Wake events that are not ignored) will result in the same `single_iteration` code running.

**Comment by rib** [comment]
Hi @notgull, can you maybe give a bit of context for closing this PR? are you maybe no longer interested in the feature?

If I were to merge the updated branch above, are you OK with keeping your attribution or should I remove it?
